airbyte-integrations/connectors/source-mysql-v2/src/main/kotlin/io/airbyte/integrations/source/mysql/MysqlJdbcPartition.kt

```diff
@@ -185,16 +185,13 @@ class MysqlJdbcSnapshotPartition(
     override val upperBound: List<JsonNode>?,
 ) : MysqlJdbcResumablePartition(selectQueryGenerator, streamState, primaryKey) {
 
+    // TODO: this needs to reflect lastRecord. Complete state needs to have last primary key value in RFR case.
     override val completeState: OpaqueStateValue
         get() =
-            when (upperBound) {
-                null -> MysqlJdbcStreamStateValue.snapshotCompleted
-                else ->
-                    MysqlJdbcStreamStateValue.snapshotCheckpoint(
-                        primaryKey = checkpointColumns,
-                        primaryKeyCheckpoint = upperBound,
-                    )
-            }
+            MysqlJdbcStreamStateValue.snapshotCheckpoint(
+                primaryKey = checkpointColumns,
+                primaryKeyCheckpoint = listOf(),
+            )
```

Comment:
@postamar I think we need to update the interface a bit to have complete state taking lastRecord into consideration. 

The ask is to have the final state message the same as the checkpoint message, at least for RFR. Use case is, if stream1 has finished but stream2 has aborted, when we resume back to RFR in the 2nd attempt, platform will not change any state but we will be able to tell there is nothing more (or there are some new records) in stream1. If we have a different value (for example, a null or 0 value) there, likely we will redo the full refresh on stream1.

What do you think? I can do that in a separate PR.

Reply:
We chatted offline - I'll create a separate new Partition, with complete state = max(pk_val), and use that for RFR. That'll be in a separate PR.